File: src/signal_parallel_handler/signal_parallel_handler.erl
-module(signal_parallel_handler).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../../docs/signal_parallel_handler.md"}.

-export([
    start/0
]).

-include("signal_parallel_handler.hrl").

-define(SIGNAL_COUNTER_WId, "signal_counter").
-define(SIGNAL_COUNTER_WE, #{workflow_id => ?SIGNAL_COUNTER_WId}).

start() ->
    temporal_sdk:start_workflow(
        ?CLUSTER,
        "default",
        signal_parallel_handler_workflow,
        [{workflow_id, ?SIGNAL_COUNTER_WId}, {workflow_execution_timeout, {1, hour}}]
    ),
    send_signals(3_000).

send_signals(Count) when is_integer(Count), Count > 0 ->
    case temporal_sdk:get_workflow_state(?CLUSTER, ?SIGNAL_COUNTER_WE) of
        {ok, running} ->
            send_signal(?PING_SIGNAL),
            case Count rem 100 of
                0 -> send_signal(?REPORT_SIGNAL);
                _ -> ok
            end,
            timer:sleep(100),
            send_signals(Count - 1);
        _ ->
            ok
    end;
send_signals(0) ->
    timer:sleep(1_000),
    send_signal(?KILL_SIGNAL).

send_signal(Signal) ->
    temporal_sdk:signal_workflow(?CLUSTER, ?SIGNAL_COUNTER_WE, Signal).

File: lib/signal_parallel_handler/signal_parallel_handler.ex
defmodule SignalParallelHandler do
  @external_resource "docs/signal_parallel_handler.md"
  @moduledoc TemporalSdk.Utils.exdoc!("docs/signal_parallel_handler.md")

  @cluster :cluster_1
  @ping_signal "ping"
  @kill_signal "kill"
  @report_signal "report"
  @signal_counter_wid "signal_counter"
  @signal_counter_we %{workflow_id: @signal_counter_wid}

  def start() do
    TemporalSdk.start_workflow(
      @cluster,
      "default",
      SignalParallelHandler.Workflow,
      workflow_id: @signal_counter_wid,
      workflow_execution_timeout: {1, :hour}
    )

    send_signals(3_000)
  end

  defp send_signals(count) when is_integer(count) and count > 0 do
    case TemporalSdk.get_workflow_state(@cluster, @signal_counter_we) do
      {:ok, :running} ->
        send_signal(@ping_signal)

        case rem(count, 100) do
          0 -> send_signal(@report_signal)
          _ -> :ok
        end

        Process.sleep(100)
        send_signals(count - 1)

      _ ->
        :ok
    end
  end

  defp send_signals(0) do
    Process.sleep(1_000)
    send_signal(@kill_signal)
  end

  defp send_signal(signal),
    do: TemporalSdk.signal_workflow(@cluster, @signal_counter_we, signal)
end
